cmd/chisel/cmd_find.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/chisel/internal/setup"
+	"github.com/canonical/chisel/internal/strdist"
+)
+
+var shortFindHelp = "Find existing slices"
+var longFindHelp = `
+The find command queries the chisel releases for available slices.
+
+With the --release flag, it queries for slices in a particular branch
+of chisel-releases repository[1] or a particular directory. If left
+unspecified, it queries with the release info found in /etc/lsb-release.
+
+[1] https://github.com/canonical/chisel-releases
+`
+
+var findDescs = map[string]string{
+	"release": "Chisel release branch or directory",
+}
+
+type cmdFind struct {
+	Release string `long:"release" value-name:"<branch|dir>"`
+
+	Positional struct {
+		Query []string `positional-arg-name:"<query>" required:"yes"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	addCommand("find", shortFindHelp, longFindHelp, func() flags.Commander { return &cmdFind{} }, findDescs, nil)
+}
+
+func (cmd *cmdFind) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	query := strings.TrimSpace(strings.Join(cmd.Positional.Query, " "))
+	if query == "" {
+		return fmt.Errorf("no search term specified")
+	}
+
+	release, releaseLabel, err := getRelease(cmd.Release)
+	if err != nil {
+		return err
+	}
+
+	slices, err := findSlices(release, query)
+	if err != nil {
+		return err
+	}
+	if slices == nil {
+		fmt.Fprintf(Stdout, "No matching slices for \"%s\"\n", query)
+		return nil
+	}
+
+	w := tabWriter()
+	fmt.Fprintf(w, "Slice\tPackage\tRelease\n")
+	for _, s := range slices {
+		fmt.Fprintf(w, "%s\t%s\t%s\n", s, s.Package, releaseLabel)
+	}
+	w.Flush()
+
+	return nil
+}
+
+const maxStrDist int64 = 3
+
+// matchSlice returns true if a slice (partially) matches with a query.
+func matchSlice(slice *setup.Slice, query string) bool {
+	// check if the query (partially) matches with "pkg_slice" format slice name.
+	dist := strdist.Distance(slice.String(), query, strdist.StandardCost, maxStrDist+1)
+	if dist <= maxStrDist {
+		return true
+	}
+	// check if the query (partially) matches with the package name.
+	distPkg := strdist.Distance(slice.Package, query, strdist.StandardCost, maxStrDist+1)
+	if distPkg <= maxStrDist {
+		return true
+	}
+	// check if the query (partially) matches with the slice name.
+	distSlice := strdist.Distance(slice.Name, query, strdist.StandardCost, maxStrDist+1)
+	if distSlice <= maxStrDist {
+		return true
+	}
+	return false
+}
```

Comment:
Why is `maxStrDist = 3`? Does this mean that when I search for `clang` I can get results for `dlang`?

Reply:
Yes, that is true. With `maxStrDist = 3`, query for `gcc` will give results for `lib` and queries for `bins` will give `libs` as well. But I think there should be scopes for partial matching. This would be particularly useful when searching for a bigger text. 

Maybe the `maxStrDist` value should be decreased when the query text is very small. What do you think?

Reply:
out of curiosity and to be consistent, have you checked what is the distance used by `snap info`?

Reply:
Do you mean `snap find`? `snap info` does not take partial matching into consideration.
As for `snap find`, yes, I did take a look. What it seems to me after following a chain of files is that it queries the store directly and gets the list of snaps from there. I could be wrong, but I did not notice any distance values used there.

Reply:
(updated :p)

right. from using it, I get the impression that the distance is 1...

Nonetheless, I'm worried about this being too big since most of our slices are typically 4 chars long (e.g. `bins`, `libs`, `data`, `core`...). Sure, if someone searches for these terms alone, we would expect a lot of results anyway, but still, querying for `bins` and getting `libs` sounds very counter intuitive. 

I'd propose starting a bit more conservatively for now and reducing that distance to 2 or even 1.

Reply:
I don't think the `snap find` command does partial matching to be honest. For example,

```sh
$ snap find chisfl
No matching snaps for "chisfl"
```
```sh
$ snap find chisl
No matching snaps for "chisl"
```

But, I think it checks if the query string is a substring (of a considerable length) of the snap name (among other things). 

```sh
$ snap find hise | grep chisel
chisel     v0.8     rebornplusplus  -      Chisel is a software tool for carving and cutting Debian packages!
```

So, we could add the following changes if you guys think it's a good idea:
- reduce the `maxStrDist` value to 2 (1 might be too less if someone is trying to find in a `python3_bins` format)
- check if the query string is a substr of the `pkg_slice` name and the query is of considerable length (maybe >50% of the slice name)

Reply:
I have this example:

```bash
$ snap find pebbe
Name    Version  Publisher   Notes    Summary
pebble  v1.3.0   canonical✓  classic  Pebble - take control of your internal daemons!
```

---

But y, let's go with your proposal (although I would drop that 50% conditional, at least for now, as it sounds like a potential improvement that we should only do once we have evidence that it is needed).

Reply:
After discussions in a call, going ahead with substring matching as @woky suggested 

> How about replacing this distance stuff with just `strings.Contains(slice.String(), query)`?

.. and a max Levenshtein distance of 1 as @cjdcordeiro  suggested.

Reply:
Changes introduced in https://github.com/canonical/chisel/pull/99/commits/3ab898e2b092df8fe2bcc37c33464b4f6a09ac26.